test/uk/gov/hmrc/agentsubscription/model/ValidatorSpec.scala

```diff
@@ -188,29 +338,226 @@ class ValidatorSpec extends UnitSpec {
+      }
+
+      "contains numbers" in {
+        validateOSEmail("123@c.d") shouldBe JsSuccess("123@c.d")
+      }
+
+      "local-part contains hyphens" in {
+        validateOSEmail("a-b@c.d") shouldBe JsSuccess("a-b@c.d")
+      }
+
+      "local-part contains full stops" in {
+        validateOSEmail("a.b@c.d") shouldBe JsSuccess("a.b@c.d")
+      }
+
+      "local-part contains +" in {
+        validateOSEmail("a+b@c.d") shouldBe JsSuccess("a+b@c.d")
+      }
+
+      "local-part contains #" in {
+        validateOSEmail("a#b@c.d") shouldBe JsSuccess("a#b@c.d")
+      }
+    }
+
+    "reject name" when {
+      "missing local-part" in {
+        validateOSEmail("@b") shouldBe emailValidationError
+      }
+
+      "missing domain" in {
+        validateOSEmail("a@") shouldBe emailValidationError
+      }
+
+      "longer than 132 characters" in {
+        val email133CharsLong = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa@b.com"
+        email133CharsLong.length shouldBe 133
+
+        validateOSEmail(email133CharsLong) shouldBe emailValidationError
+      }
+    }
   }
 
-  private def validateName(name: String) = {
-    nameValidation.reads(JsString(name))
+  "safeIdValidation" should {
+    def validateSafeId(safeId: String) = safeIdValidation.reads(JsString(safeId))
+    def safeIdValidationError = JsError(ValidationError("error.safeid.invalid"))
+
+    "accept SafeId" when {
+      "it's a perfectly normal SafeID" in {
+        val validSafeId = "XE0001234567890"
+        validateSafeId(validSafeId) shouldBe JsSuccess(validSafeId)
+      }
+
+      "second character is any uppercase letter" in {
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZ".foreach { secondCharacter =>
+          val validSafeId = s"X${secondCharacter}0001234567890"
+          validateSafeId(validSafeId) shouldBe JsSuccess(validSafeId)
+        }
+      }
+    }
+
+    "reject SafeId" when {
+      "contains lowercase characters" in {
+        val invalidSafeId = "xe0001234567890"
+        validateSafeId(invalidSafeId) shouldBe safeIdValidationError
+      }
+
+      "first character is not 'X'" in {
+        val invalidSafeId = "YE0001234567890"
+        validateSafeId(invalidSafeId) shouldBe safeIdValidationError
+      }
+
+      "second character is not a letter" in {
+        val invalidSafeId = "X00001234567890"
+        validateSafeId(invalidSafeId) shouldBe safeIdValidationError
+      }
+
+      "3rd - 5th characters are not '000'" in {
+        val validSafeIds = for {
+          third <- 0 to 9
+          fourth <- 0 to 9
+          fifth <- 0 to 9
+          if (0 != (third + fourth + fifth))
+        } yield s"X$third$fourth${fifth}1234567890"
+
+        validSafeIds.foreach { validSafeId =>
+          validateSafeId(validSafeId) shouldBe safeIdValidationError
+        }
+      }
+
+      "6th through 15th character is not 0-9" in {
```

Comment:
6th through 15th?? a bit confusing, can you make this clear

Reply:
Sure, hopefully clearer now.